Add Image Compare block

mkaz opened the pull request:
Adds the Image Compare block as an experimental block, the image compare blocks allows you to place two images side-by-side (or above/below)  and use a slider to reveal different aspects of them.

Here's an example:

![image-compare](https://user-images.githubusercontent.com/45363/80538575-d0b7f480-895a-11ea-85b8-9221e536d30f.gif)

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
Internal reference: p1HpG7-8NH-p2

#### Testing instructions:

* This is a beta block, so you will need to set  in wp-config.php
    `define( 'JETPACK_BETA_BLOCKS' true );`

* Create new post . Add block to post and select before/after images
* The comparison tool should show within the editor, as it will look on front-end
* Confirm resizing editor the images resize as expected
* Confirm shows on front-end and works as expected
* Test various cases of missing images or unexpected usage

#### Proposed changelog entry for your changes:

* Adds new Image Compare block 

## Future Tracking Issues;

* What to show for AMP pages that can't load library? (See: #15653)
* Fix useResizeObserver to support older (current) WordPress versions without the Gutenberg plugin (See: #15655)

## Licensing

**NOTE:** This block uses code originally from the Knight Labs Juxtapose block, I have discussed with @pesieminski and he has cleared the license notice in the readme.md and at the top of the view.js files.

Files here:
https://github.com/Automattic/jetpack/pull/15598/files#diff-6bee5a8a6f949abff285c30df8232b6e
https://github.com/Automattic/jetpack/pull/15598/files#diff-84156f4fb84990fe9b2aba4b50eef7a5

jasmussen commented:
Wonderful, Marcus! Excited about this block.

> What to show for AMP pages that can't load library?

This is a common problem with blocks that feature interactivity, and as I noted in [this ticket I created](https://github.com/WordPress/gutenberg/issues/17764), I would love for there to be a generic fallback for blocks that aren't able to render in AMP contexts.

Specifically, just providing a button that says "View on [sitename]" feels like a fine solution, not just generically but for this block as well.

So the bigger questions are of a technical nature:

1. Are we able to detect the AMP context?
2. If yes, are we able to detect e-mail or other lo-fi contexts?
3. If not, perhaps we simply need to output this button if the JavaScript doesn't load.

If we were able to answer "yes" to 1 and 2, not only should we consider submitting this AMP & email fallback solution to the block editor project itself so all complex blocks could benefit from it, but we would be able to build in the button that says "View on [site]" and have confidence that clicking the button would work.

The problem with 3 is that it would also invoke in contexts where the user has disabled JavaScript, and begs the question whether we therefore also need to output text that suggests JavaScript is necessary.

Still, 3 seems the most straight-forward, an in absence of 1 and 2, we could do this:

```
<noscript>
<div class="wp-block-button">
	<a class="wp-block-button__link" href="[url]">View on [sitename]</a>
</div>
</noscript>
```

I'd tend to lean towards not needing the JS warning. Virtually everything requires JS these days, so if you actively disable it, you'll likely have an idea why things might not be working.

jasmussen commented:
Great thoughts, Jeremy, I'll respond to the conceptual and design aspects and let Marcus help with the difficult stuff (thanks Marcus).

> When switching from side by side to above and below, should the placeholder be updated accordingly?

Not a bad idea. 

It's not trivial to do, though, and the Placeholder component is a little in flux at the moment. Would you mind if I file this for a follow-up?

> What if we used <amp-carousel> for now (reference)? It doesn't have the same fancy effect, but it would display the 2 images next to each other.

I'm actually concerned about providing an experience that isn't identical to what the user asked for. The fanciness of this block is exactly that it stacks the two images on top of each other and let's you interactively scrub the separator to see the difference. This is what you see in the editor and anything else has the potential to ruin the effect.

Imagine someone using this for a fun little "drag to reveal spoiler" and amp then reveals it ahead of time?

Or, more likely, the user is likely to write in the prose: "compare the images below, drag the slider to reveal", which will obviously be misleading in an amp context.

For that reason, I'd prefer a "view on site" button.

jeherve commented:
That's a good point, whatever workaround we'd pick, it'd be quite far from the original behaviour of the block.

I don't know if we can implement a "view on site" button though; while that could be an option with sites using AMP's transitional mode, I don't think it would be possible on sites using AMP's standard mode, where the site is AMP-first.

cc @westonruter; would you have some ideas?

mkaz commented on extensions/index.json:
```diff
@@ -31,7 +31,7 @@
 	],
 	"beta": [ "amazon", "instagram-gallery" ],
-	"experimental": [ "seo" ],
+	"experimental": [ "image-compare", "seo" ],
```
👍 I'll switch it to beta then

mkaz commented on extensions/blocks/image-compare/use-debounce.js:
```diff
@@ -0,0 +1,13 @@
+ */
+import { useEffect } from '@wordpress/element';
+
+const useDebounce = ( callback, delay, deps ) => {
```
I submitted it to core as a new hook. PR is here: https://github.com/WordPress/gutenberg/pull/21179

simison replied:
I've quite enjoyed external package for this to avoid any maintenance hurdles: https://www.npmjs.com/package/use-debounce

jasmussen commented:
> I don't know if we can implement a "view on site" button though; while that could be an option with sites using AMP's transitional mode, I don't think it would be possible on sites using AMP's standard mode, where the site is AMP-first.

What does CNN do for their PDF reader which doesn't work on AMP? It's part of the screenshot shown in https://github.com/WordPress/gutenberg/issues/17764

jeherve commented:
> What does CNN do for their PDF reader which doesn't work on AMP?

I'll let @westonruter weigh in on that, but I believe that's only an option if your site has two different versions available: AMP pages and your native site. When using AMP's standard mode in the WordPress plugin, your native site already uses AMP. Here is an example:
https://medinathoughts.com/2019/04/30/amp-in-wordpress-the-wordpress-way/
The page above is served via AMP out of the box, there is no alternative URL.

jasmussen commented:
Understood — but doesn't AMP entirely strip out external JavaScript, making it possible to serve the button using `noscript` tags? I will obviously defer to the experts here, as I haven't brushed up my AMP skills lately.

westonruter commented:
Note that AMP has an [`amp-image-slider`](https://amp.dev/documentation/components/amp-image-slider/) component for this very use case. So there should be no need to gracefully degrade here.

The only thing the AMP component isn't supporting yet is the ability to compare vertically. There is an issue for this but it just needs to get bumped I think: https://github.com/ampproject/amphtml/issues/18634.

Until that is resolved, one option could be to hide that orientation toggle when the AMP plugin is active.

westonruter commented on extensions/blocks/image-compare/image-compare.php:
```diff
@@ -0,0 +1,60 @@
+ * @return string
+ */
+function render_amp( $attr ) {
+	// @TODO: what to show for AMP view?
+	// :shrug-emoji:
```
As noted in https://github.com/Automattic/jetpack/pull/15598#issuecomment-621372376, use the [`amp-image-slider`](https://amp.dev/documentation/components/amp-image-slider/) component.

mkaz replied:
Ticket opened in #15653 to address

nainar commented:
What kind of timeline would you need support for vertical image slider to be implemented in AMP?

mkaz commented on extensions/blocks/image-compare/style.scss:
```diff
@@ -0,0 +1,9 @@
+.wp-block-jetpack-image-compare-block {
```
Removed.

mkaz commented:
> What kind of timeline would you need support for vertical image slider to be implemented in AMP?

@nainar Ideally would be in the next couple weeks, our target is the June Jetpack release, with feature freeze end of May — but if it takes longer we might be able to code around it as Weston suggested, or just not have that feature available in AMP until it's ready. Thanks!

jasmussen commented:
That's a good point, if the AMP image slider does not support vertical comparison, that definitely enables users to get into situations where what they publish in AMP format is not what they expected.

A button, "View on [sitename]" would be a somewhat dull fallback for that, but at least it would work in the interim.

Alternately we can accept that it doesn't work for the time being, and revisit when it does?

A third option could be to detect the presence of the AMP plugin and simply disable the option for vertical comparison.

I'd personally be fine with either of the above approaches, but I would lean towards the **simplest** solution that gets us shipping. I would rather we ship a small thing and learn from it, than assume that this block will set the world on fire and optimize it to perfection only to learn it gets few downloads.

jasmussen commented:
Just spoke with Jeremy about some particular use cases of AMP, and was impressed to learn that an entire website could be AMP-only!

In such situations, of course, there wouldn’t be anywhere but that website to see the contents of the block, so a “view on [sitename]” button as referenced from the CNN example does not make sense. 

This leaves us a couple of options to move forward with:

1. Wait for vertical comparison to land in the AMP slider, and leverage that here. 
2. Hide the option to choose vertical comparison in AMP contexts, until it lands.
3. Move forward without AMP support for the time being, and add it back when possible.
What do you think?

We have a month to go before a Jetpack code freeze, so we might be able to leverage 1. But if not, both 2 and three seem like good options to keep the momentum going. 

In any case, it doesn't all have to happen in one PR, so we might want to land this PR without amp support, and then look at followups.

mkaz commented:
@jasmussen I'm going to aim for (2) that is if I can detect the plugin then hide the vertical option.

westonruter commented:
> I'm going to aim for (2) that is if I can detect the plugin then hide the vertical option.

@mkaz An easy way to detect is to check if the `is_amp_endpoint` function exists.

jeherve commented:
> An easy way to detect is to check if the is_amp_endpoint() function exists.

We have a helper for that, it may be useful:
`class_exists( 'Jetpack_AMP_Support' ) && Jetpack_AMP_Support::is_amp_request()`

You can check how other blocks use it here, for example:
https://github.com/Automattic/jetpack/blob/70dab921662987d1dd1aa7e0f4f302dc692989d2/extensions/blocks/google-calendar/google-calendar.php#L52-L66

westonruter commented:
> We have a helper for that, it may be useful:
> `class_exists( 'Jetpack_AMP_Support' ) && Jetpack_AMP_Support::is_amp_request()`

That's not quite what is needed here. The question is not whether it is an AMP response but whether the AMP plugin is installed and active at all. If it is, then the vertical option should be hidden since it's not quite available in AMP yet.

mkaz commented:
@jeherve I've rebased, and fixed up the Travis tests, plus I've created tracking issues for AMP (#15653) and older editor support (#15654) 

I think its ready for review, when you get the chance.  👍 

Thanks!

mkaz commented on extensions/blocks/image-compare/edit.js:
```diff
@@ -0,0 +1,146 @@
+ */
+import { InspectorControls, RichText } from '@wordpress/block-editor';
+import { PanelBody, RadioControl, Placeholder } from '@wordpress/components';
+import { useResizeObserver } from '@wordpress/compose';
```
Ticket opened here to addresss; #15654

mkaz commented on extensions/blocks/image-compare/edit.js:
```diff
@@ -0,0 +1,147 @@
+	);
+
+	return (
+		<figure className="wp-block-jetpack-image-compare-block" id={ clientId }>
```
👍

mkaz commented on extensions/blocks/image-compare/save.js:
```diff
@@ -0,0 +1,41 @@
+	} = attributes;
+
+	return (
+		<figure>
```
👍

mkaz commented:
@jeherve ok, I was able to duplicate, it is only if you upload and it takes awhile. Also, I'm not sure why it is top-bottom on upload and then switches to side-by-side. 

I'll look to see what we can do for the upload indicator.

mkaz commented:
@jeherve Updated with:

1. Error checking - catches if the files are too big, or if the API returns an error
2 WP Compatibility - added check for `useResizeObserver` so supports WP 5.4 and older

I still need the loading indicator for when the file is being uploaded, the temporary/blob image is in place, so there isn't a visual indicator to the user that it is not really uploaded. 
@jasmussen any thoughts?

mkaz commented:
ok, after a bit of investigation I think the placeholder is not showing due to having two `MediaPlaceholder`s in the component at once. Reading through how the core image blocks works, it looks like the upload state is handled by the `withSelect` data that returns a single prop, likewise with the `withNotices` and error messages.

So I think we need to rework a bit, I believe our options our:

1. Use InnerBlocks for the two images, then each can be its own component

2. Use multiple settings for uploader so user can select both images at once, this might introduce more usability issues by being able to select more than one, selecting/deselecting might confuse the ordering, etc...

3. Other idea?

jeherve commented:
> Use InnerBlocks for the two images, then each can be its own component

I like this, that seems like a good idea!

simison commented on extensions/blocks/image-compare/index.js:
```diff
@@ -0,0 +1,86 @@
+		attributes: {
+			imageBeforeId: '1',
+			imageBeforeUrl:
+				'https://upload.wikimedia.org/wikipedia/commons/archive/d/d8/20180325051241%21Lynda_Carter_Wonder_Woman.JPG',
```
I think we'd wanna refer to an image locally instead of Wikimedia? Other image blocks in jetpack have examples.

mkaz replied:
Updated, didn't realize you could do that in Jetpack 👍

simison commented on extensions/blocks/image-compare/index.js:
```diff
@@ -0,0 +1,86 @@
+
+	icon,
+
+	category: 'layout',
```
This would need fallback support for jetpack category, example:

https://github.com/Automattic/jetpack/blob/21daeebcab1798413b86f85bae01f3cc9f53ad92/extensions/blocks/business-hours/index.js#L86

mkaz replied:
Done. 👍

simison commented on extensions/blocks/image-compare/view.scss:
```diff
@@ -0,0 +1,337 @@
+/**
+ * Juxtapose - v1.2.0 - 2017-12-18
```
Latest is from 2019 1.2.1:
 — if to fork now here, why not take the latest? 

https://github.com/NUKnightLab/juxtapose/blob/94da426b3906f1d305a89d115f8fa3da67827d0c/build/js/juxtapose.js#L1

mkaz replied:
The last commit is adding Google Tag Manager support: https://github.com/NUKnightLab/juxtapose/commits/94da426b3906f1d305a89d115f8fa3da67827d0c/build/js/juxtapose.js

So we're not missing anything, and not what we want. There has also been a fair amount of modification to the library, to modernize and remove features such as a bunch of weird built-in Flickr stuff.

mkaz commented:
@jeherve I switched out to use `InnerBlocks` and I think it is better, code is a bit simpler and works with regards to upload/error handling for each image. Test it out and let me know if it works for you like this.

@jasmussen Can you take a look at the design, in particular the layout of the side-by-side images on upload. I had to modify the CSS since the markup changed, and gave it a good shot, but still probably can use a little more CSS ❤️

mkaz commented on extensions/blocks/image-compare/editor.scss:
```diff
@@ -0,0 +1,90 @@
+
+.image-compare__error {
+	font-size: 14px;
+	color: red;
```
This was a temporary thing, we worked to use InnerBlocks and standard notices.

mkaz commented on extensions/blocks/image-compare/editor.scss:
```diff
@@ -0,0 +1,90 @@
+	position: relative;
+}
+
+.image-compare__error {
```
This was a temporary thing, we worked to use InnerBlocks and standard notices.

mkaz commented on extensions/blocks/image-compare/edit.js:
```diff
@@ -0,0 +1,211 @@
+								</div>
+								<MediaPlaceholder
+									onError={ err => {
+										setErrorMessage( `Error uploading: ${ err[ 2 ] }` );
```
This was a temporary thing, reworked to use InnerBlocks and standard notices.

mkaz commented on extensions/blocks/image-compare/index.js:
```diff
@@ -0,0 +1,86 @@
+/**
+ * WordPress dependencies
```
I'll update, but this probably should be reconsidered. WordPress dependencies is a bit different than External, but really doesn't make much difference.

mkaz commented on extensions/blocks/image-compare/edit.js:
```diff
@@ -0,0 +1,211 @@
+				} );
+			}
+		}
+	} );
```
Reworked block, so this no longer applies.

mkaz commented:
> How did other alternatives to Juxtapose library look like? It's very unfortunate we need to copy it here instead of relying on npm. That basically guarantees it's never gonna get updated. :-/

I consider this a good thing. It is rather simple and works as expected, it hasn't been updated in many years and yet persists. We can apply the [Lindy effect](https://en.wikipedia.org/wiki/Lindy_effect) a theory about that the future life expectancy of a technology is proportional to its current age. So the fact that it has been around a long time and not changing is good, it means it will be around much longer than something that is new and rapidly changing.

jasmussen commented:
Thank you all for keeping on this! 

We were so close with the block, and then this issue occurred which required a refactor. It's always like this, isn't it?

Here's what i see now:

![image compare](https://user-images.githubusercontent.com/1204802/81278163-9be02780-9055-11ea-9d88-85c2f7629930.gif)

The block now uses two innerblocks, one for each image, locked in place. Why the end result doesn't work, I sense that was an issue with my local setup, it took a while to get jetpack up and running fresh, I may have missed a few steps.

However from just the inital setup experience, unfortunately I find it substantially worse:

- What is essentially one block that compares two images, is now three blocks — the parent, and two children.
- Each child can be selected and receive focus with the mouse, but tabbing through the blocks only tabs through the first one. Not only that, but with the addition of innerblocks, there are numerous additional tabstops.
- Each child remains a child block, which means adjusting things like the parent block's alignment requires selecting that layer first.
- You can't just select the block and press delete to remove it, which you could before.

There is absolutely a place for blocks with nesting, and those blocks come with their own design challenges and constraints. But this block was designed to be a single block, and I do not feel comfortable shipping it as is. We need to rethink this approach, I'm afraid.

Can we take inspiration from Media & Text, where the placeholder on the left is essentially a child block, but is otherwise not something we can interact with?

![media and text](https://user-images.githubusercontent.com/1204802/81278889-9b945c00-9056-11ea-88e4-f3ae48cad3f8.gif)

To frame it differently, in order for this block to be a good experience for users, it has to appear as a single block with a single selection target.

simison commented:
Does [`__experimentalcapturetoolbars`](https://github.com/WordPress/gutenberg/blob/52755706287eb6bf90e3abf965cc40b9fecb3444/packages/block-editor/src/components/inner-blocks/README.md#__experimentalcapturetoolbars) change select dynamics somehow that it would work here?

jasmussen commented:
Good instinct, but that would still not let you click a placeholder and press Delete to remove it, and it would come with the same amount of tabstops.

There has to be a way to put two upload dropzones next to each other without having two blocks. 

What does media & text do for the left Placeholder to make it appear as part of the parent block?

mkaz commented:
Yet another attempt, this one actually feels much better and should've tried this before InnerBlocks but live and learn. Basically, I create a `ImgUpload` component and attach `withNotices` to that and then include the component twice for each image. Works as expected (for me)

Likely needs a little CSS polish that I might of lost during the iterations but should behave more appropriately as a single block, less markup and selection targets, and tabbing works as expected.

jasmussen commented:
This is what I see. 

![image compare v2](https://user-images.githubusercontent.com/1204802/81329408-d02c0600-909e-11ea-958e-2258561fb24e.gif)

First the good news: this is solid. Outside of a few CSS changes that I'll help make, this is exactly the experience I was hoping for. You even got drag and drop to work, and I also managed to invoke an error message (too large filesize) and all was handled well. I can select a single block, press tab 4 times to get to the 4th button inside, press delete to remove the block.

:chef-kisses-fingers:

Well done. 

The bad news as you can see at the end of the GIF, I still can't get the dang thing to render — I'm not sure why exactly, I still think it can be because of my setup of the jetpack plugin — I cloned this repo into my gutenberg install, so it's running latest master of the Gutenberg plugin. But it doesn't make a difference if I deactivate the Gutenberg plugin. I can inspect the block and it looks like the ingredients are there, just that some JS didn't fire correctly. I see the following warning on the frontend:

<img width="1097" alt="Screenshot 2020-05-07 at 20 09 58" src="https://user-images.githubusercontent.com/1204802/81329791-55171f80-909f-11ea-91f9-26ccd7c762ab.png">

I don't think it's related, but thought I'd share.

But again — the experience is great. I'll push some CSS tomorrow, and hopefully we can get this going! Stellar work Marcus!

mkaz commented:
I think the render issue was due to the blob URL, which is the temporary image that it shows when uploading. I switched this to match how the Media & Text block does it, which is not to show the image until uploaded. This behaves more consistently, and less error prone.

mkaz commented:
In testing, I noticed another issue was using images of different sizes or aspect ratios. There was an aggressive check in the library that would bail if they weren't the same. I removed this check since, it all still works, though mixing landscape and portrait gives interesting results, but no reason to break because of it.

jasmussen commented:
Happy to report that the block works for me now:

<img width="733" alt="Screenshot 2020-05-08 at 07 55 06" src="https://user-images.githubusercontent.com/1204802/81375186-406b7380-9101-11ea-9f97-58e4cfd435d2.png">


The issue was almost certainly the check for dimensions. I was a little haphazard in my testing and used somewhat random things. But agree it's good to remove those checks.

Will push a little CSS now. Thanks Marcus!

jasmussen commented:
Okay, I pushed some polish. There was some cruft to remove, some placeholders to adjust, some typos to fix, and some CSS to polish up. I think I caught all of it, and it now feels ready for final review again:

![image compare v3](https://user-images.githubusercontent.com/1204802/81377591-25e7c900-9106-11ea-8665-dc151352126f.gif)

jeherve commented on extensions/blocks/image-compare/view.js:
```diff
@@ -0,0 +1,559 @@
+			this.imgAfter = new Graphic( images[ 1 ], this );
+		}
+
+		if ( this.imgBefore.credit || this.imgAfter.credit ) {
+			this.options.showCredits = true;
+		} else {
+			this.options.showCredits = false;
+		}
+	}
```
Are credits used anywhere in our implementation, or is this something we could get rid of?

mkaz replied:
Good catch, removed.

jeherve commented on extensions/blocks/image-compare/view.js:
```diff
@@ -0,0 +1,559 @@
+		return dimensions;
+	}
+
+	function addClass( element, c ) {
+		if ( element.classList ) {
+			element.classList.add( c );
+		} else {
+			element.className += ' ' + c;
+		}
+	}
```
I didn't experience any issues with this in Jetpack, but ran into the following error in the editor, after uploading the first image in the block, when testing D42965-code:

```
Uncaught TypeError: Cannot read property 'classList' of null
```

jeherve replied:
Logged here: #15743

mkaz commented:
> As we discussed earlier, I find the lack of progress indicators disturbing. It's hard to know that the block is really working when you upload images.

Agree, I think this can be a nice feature to add. Right now it matches the core "Media & Text" block, the current way to add an indicator is a bit complex. I plan to write up a ticket for improving media upload experience.

> Once you've uploaded images, you cannot edit / replace one or both of them.

This is known, initially this was considered an added layer of user interface complexity and that it would be easier to just delete the block and start over then add in a complex click sequence to select to replace and not interfere with the slider. However, we might be able to do something with the new Replace Image toolbar and G2 interfaces. @jasmussen thoughts?

> As I mentioned in my last review, I think it'd be nice if we could leverage our existing image tag creation tools; it would give us more data per image, such as srcset.

We can test some things out, but the library reads the main src attribute, so not sure it would benefit having a srcset of images.

> I think it would be nice if the block supported wide and full sizes, since that can be handy for images we want to display.

Good suggestion, we should add this feature.

jasmussen commented:
Such good reviews, such good conversation. You're all wonderful.

> As we discussed earlier, I find the lack of progress indicators disturbing. It's hard to know that the block is really working when you upload images.

Agreed — for V2. But we might even want to look at making that PR to the core project, if we can somehow benefit Media & Text, then perhaps we would not have to write one-off code for this block, and benefit both.

> Once you've uploaded images, you cannot edit / replace one or both of them.

Agreed — for V2.

> We can test some things out, but the library reads the main src attribute, so not sure it would benefit having a srcset of images.

Seems fine to explore, but I'd go at least V2, possibly even V3.

> I think it would be nice if the block supported wide and full sizes, since that can be handy for images we want to display.

Also agreed, but also V2, or at the very least a followup PR. This one may seem trivial, but I have that tingling sensation that it is absolutely not at all. Already now we had some challenges with the responsiveness of the block, I imagine the hooks we have to fire to re-fire the JS will take more time to get right than we think it does.

I truly don't mean to be dismissive of any review points by suggesting features be V2 or V3, but rather just that we have to remember that this is an entirely new block, and although we assume people will love it, but we might be looking at an extremely niche block used once or twice in 10.000 posts. 

By putting out a fully functional and nice version, as this PR has, we'll keep things moving and can prioritize our resources where they benefit the most. 

And I'm not just saying this because I need Marcus' help on our V2 of the Timeline block ;)

jeherve commented:
r207236-wpcom
